### ISSUE TITLE
A potential goleak in libnetwork/networkdb/cluster.go

### ISSUE THREAD

**Issue by xuxiaofan1203** [moby/moby#47859]
### Description

Hello, I found in the function clusterLeave()https://github.com/moby/moby/blob/ceefb7d0b9c5b6fbd1ea7511592a4ddb28ec4821/libnetwork/networkdb/cluster.go#L222

```
func (nDB *NetworkDB) clusterLeave() error {
	mlist := nDB.memberlist


	if err := nDB.sendNodeEvent(NodeEventTypeLeave); err != nil {
		log.G(context.TODO()).Errorf("failed to send node leave: %v", err)
	}


	if err := mlist.Leave(time.Second); err != nil {
		return err
	}


	// cancel the context
	nDB.cancelCtx()


	for _, t := range nDB.tickers {
		t.Stop()
	}


	return mlist.Shutdown()
}
```
If the mlist.Leave() return err, the nDB.cancelCtx() below will not get executed.
https://github.com/moby/moby/blob/ceefb7d0b9c5b6fbd1ea7511592a4ddb28ec4821/libnetwork/networkdb/cluster.go#L229-L234

And it will lead the <-nDB.ctx.Done in triggerFunc() blocked persistently, so the goroutine leak.
https://github.com/moby/moby/blob/ceefb7d0b9c5b6fbd1ea7511592a4ddb28ec4821/libnetwork/networkdb/cluster.go#L176

blocking position:https://github.com/moby/moby/blob/ceefb7d0b9c5b6fbd1ea7511592a4ddb28ec4821/libnetwork/networkdb/cluster.go#L251-L258

### Reproduce

I reproduce the bug by goleak.
Firstly, I modified the judge condition from err != nil to err == nil. Because I don't know how to let err != nil, the change only to make the return err can be executed easier. I'm not sure whether the change can lead other influences.
Normally:
```
	if err := mlist.Leave(time.Second); err != nil {
		return err
	}
```
After modified:
```
	if err := mlist.Leave(time.Second); err == nil {
		return err
	}
```
Then I used goleak to test in these test function related the funciton.
https://github.com/moby/moby/blob/ceefb7d0b9c5b6fbd1ea7511592a4ddb28ec4821/libnetwork/networkdb/networkdb_test.go#L180
Like this:
![O` O$Y@3FVFWM%V4H3AQ8EH](https://github.com/moby/moby/assets/61485003/7f4aea21-d33d-44b1-b4f8-fb201f41803a)
The result shows that there is a bug at the  <-nDB.ctx.Done
![DPNHW@V DSD))B4B5~TO5M](https://github.com/moby/moby/assets/61485003/ca08f174-fefa-4735-be3e-fe4839b4ae6c)

### docker version

```bash
latest
```

### docker info

```bash
latest
```

### Additional Info

In short, I think the bug is caused by return but have not called the cancelFunc. I have tried to describe it in detail.

**Comment by thaJeztah** [comment]
cc @akerouanton @corhere
